### ISSUE TITLE
Form Validation: Array Required

### ISSUE THREAD

**Issue by splodgeco** [bcit-ci/CodeIgniter#3772]
In 2.2 I used the standard required check to ensure the received post data included a field.  Sometimes the field was an array and as a result the form validation would fail if the array was not passed.

Since the update to version 3, the required check in regards to an array, checks within the array and fails if one of the keys (even in multidimensional arrays) does not have a value.  This might not be a bug and is expected behavior now however I used this simple check to ensure the array was transmitted with the post data.  I use a callback method with the array data so it's not to difficult to remove the "required" check into my callback.

Just thought I would let you know in case it was a bug or it's behavior is unexpected.

Cheers.

**Comment by narfbg** [comment]
Again, please give an example ... it's really hard to follow your explanation.

**Comment by Dentxinho** [comment]
I have the same problem here and I was investigating it right now.

Example:
Checkboxes (or inputs, whatever) on the form with names:

* field[foo][bar]
* field[foo][baz]

In 2.2, if we set a 'required' rule on 'foo' (`$this->form_validation->set_rules('foo', 'Foo', 'required');`, `run()` returns `TRUE` if `field[foo][bar]` is not empty OR `field[foo][baz]` is not empty. In 3.x, it returns `FALSE` on the same conditions, and both fields must not be empty to get a `TRUE`.

@narfbg it might be a bug on the recursive calls to `_execute` AND/OR wrong treatment of `$row['is_array']`.
On my example, on the most inner recursive calls, it sets `NULL` to `$postdata` (Form_validation.php Line 696) then messes field postdata on line 767. After that, required fails (postdata is now null)

**Comment by splodgeco** [comment]
That's what I meant and a much better explanation.

**Comment by Dentxinho** [comment]
Just edited my comment (i meant 'field' with a required rule, not 'foo')

**Comment by Dentxinho** [comment]
update:

Changing rule to 'field[]' (`$this->form_validation->set_rules('field[]', 'Field', 'required')`) does the trick.

**Comment by narfbg** [comment]
Yes, you need to be verbose about it ... If you write 'field' but then an array is submitted, it should mean that you intended to receive a string, but somebody has modified the form - a failure is completely natural in that case.
